master/internal/api_trials.go

```diff
@@ -50,11 +51,36 @@ var (
 
 	// TrialAvailableSeriesBatchWaitTime is exported to be changed by tests.
 	TrialAvailableSeriesBatchWaitTime = 15 * time.Second
-
-	// Common errors.
-	trialNotFound = status.Error(codes.NotFound, "trial not found")
 )
 
+func (a *apiServer) canGetTrialsExperimentAndCheckCanDoAction(ctx context.Context,
+	trialID int, actionFunc func(model.User, *model.Experiment) error,
+) error {
+	curUser, _, err := grpcutil.GetUser(ctx, a.m.db, &a.m.config.InternalConfig.ExternalSessions)
+	if err != nil {
+		return err
+	}
+
+	trialNotFound := status.Errorf(codes.NotFound, "trial %d not found", trialID)
+	exp, err := a.m.db.ExperimentWithoutConfigByTrialID(trialID)
+	if errors.Is(err, db.ErrNotFound) {
```

Comment:
hmm I think that is interesting, I wonder if we could somehow use some sort of middleware instead.

The defer setting a variable feels a little magic to me and you still have some boilerplate of the ```defer func()``` and the need to set it.

Either way I think I am going to revisit this once we start implementing the interface to see how we return errors since that ```mapAndFilter``` will put a requirement on the authentication errors we return.